src/atomic_c.c4

```diff
@@ -54,107 +64,209 @@ define(`SHMEM_PROF_DEF_CSWAP',
 SHMEM_DEFINE_FOR_BITWISE_AMO(`SHMEM_PROF_DEF_FETCH_AND')
 
+define(`SHMEM_PROF_DEF_CTX_FETCH_AND',
+`#pragma weak shmem_ctx_$1_atomic_fetch_and = pshmem_ctx_$1_atomic_fetch_and
+#define shmem_ctx_$1_atomic_fetch_and pshmem_ctx_$1_atomic_fetch_and ')dnl
+SHMEM_DEFINE_FOR_BITWISE_AMO(`SHMEM_PROF_DEF_CTX_FETCH_AND')
+
 define(`SHMEM_PROF_DEF_OR',
 `#pragma weak shmem_$1_atomic_or = pshmem_$1_atomic_or
 #define shmem_$1_atomic_or pshmem_$1_atomic_or')dnl
 SHMEM_DEFINE_FOR_BITWISE_AMO(`SHMEM_PROF_DEF_OR')
 
+define(`SHMEM_PROF_DEF_CTX_OR',
+`#pragma weak shmem_ctx_$1_atomic_or = pshmem_ctx_$1_atomic_or
+#define shmem_ctx_$1_atomic_or pshmem_ctx_$1_atomic_or')dnl
+SHMEM_DEFINE_FOR_BITWISE_AMO(`SHMEM_PROF_DEF_CTX_OR')
+
 define(`SHMEM_PROF_DEF_FETCH_OR',
 `#pragma weak shmem_$1_atomic_fetch_or = pshmem_$1_atomic_fetch_or
 #define shmem_$1_atomic_fetch_or pshmem_$1_atomic_fetch_or')dnl
 SHMEM_DEFINE_FOR_BITWISE_AMO(`SHMEM_PROF_DEF_FETCH_OR')
 
+define(`SHMEM_PROF_DEF_CTX_FETCH_OR',
+`#pragma weak shmem_ctx_$1_atomic_fetch_or = pshmem_ctx_$1_atomic_fetch_or
+#define shmem_ctx_$1_atomic_fetch_or pshmem_ctx_$1_atomic_fetch_or')dnl
+SHMEM_DEFINE_FOR_BITWISE_AMO(`SHMEM_PROF_DEF_CTX_FETCH_OR')
+
 define(`SHMEM_PROF_DEF_XOR',
 `#pragma weak shmem_$1_atomic_xor = pshmem_$1_atomic_xor
 #define shmem_$1_atomic_xor pshmem_$1_atomic_xor')dnl
 SHMEM_DEFINE_FOR_BITWISE_AMO(`SHMEM_PROF_DEF_XOR')
 
+define(`SHMEM_PROF_DEF_CTX_XOR',
+`#pragma weak shmem_ctx_$1_atomic_xor = pshmem_ctx_$1_atomic_xor
+#define shmem_ctx_$1_atomic_xor pshmem_ctx_$1_atomic_xor')dnl
+SHMEM_DEFINE_FOR_BITWISE_AMO(`SHMEM_PROF_DEF_CTX_XOR')
+
 define(`SHMEM_PROF_DEF_FETCH_XOR',
 `#pragma weak shmem_$1_atomic_fetch_xor = pshmem_$1_atomic_fetch_xor
 #define shmem_$1_atomic_fetch_xor pshmem_$1_atomic_fetch_xor')dnl
 SHMEM_DEFINE_FOR_BITWISE_AMO(`SHMEM_PROF_DEF_FETCH_XOR')
 
+define(`SHMEM_PROF_DEF_CTX_FETCH_XOR',
+`#pragma weak shmem_ctx_$1_atomic_fetch_xor = pshmem_ctx_$1_atomic_fetch_xor
+#define shmem_ctx_$1_atomic_fetch_xor pshmem_ctx_$1_atomic_fetch_xor')dnl
+SHMEM_DEFINE_FOR_BITWISE_AMO(`SHMEM_PROF_DEF_CTX_FETCH_XOR')
+
 define(`SHMEM_PROF_DEF_ATOMIC_SET',
 `#pragma weak shmem_$1_atomic_set = pshmem_$1_atomic_set
 #define shmem_$1_atomic_set pshmem_$1_atomic_set')dnl
 SHMEM_DEFINE_FOR_EXTENDED_AMO(`SHMEM_PROF_DEF_ATOMIC_SET')
 
+define(`SHMEM_PROF_DEF_CTX_ATOMIC_SET',
+`#pragma weak shmem_ctx_$1_atomic_set = pshmem_ctx_$1_atomic_set
+#define shmem_ctx_$1_atomic_set pshmem_ctx_$1_atomic_set')dnl
+SHMEM_DEFINE_FOR_EXTENDED_AMO(`SHMEM_PROF_DEF_CTX_ATOMIC_SET')
+
 #endif /* ENABLE_PROFILING */
 
+
+
 #define SHMEM_DEF_SWAP(STYPE,TYPE,ITYPE)                        \
     TYPE SHMEM_FUNCTION_ATTRIBUTES                              \
-    shmem_##STYPE##_swap(TYPE *target, TYPE value, int pe)      \
-    {                                                           \
+    SHMEM_FUNC_PROTOTYPE(STYPE##_swap, TYPE *target,            \
+                         TYPE value, int pe)                    \
```

Comment:
I _think_ you we merge the _atomic* generators if you move the STYPE and op pasting into the prototype macro (e.g. invoke as ```SHMEM_FUNC_PROTOTYPE(STYPE, swap, ...)```) and define:

```#define SHMEM_FUNC_PROTOTYPE(TYPE, OP, ...) shmem_##TYPE##_##OP(__VA_ARGS__)```
```#define SHMEM_FUNC_PROTOTYPE(TYPE, OP, ...) shmem_##TYPE##_atomic_##OP(__VA_ARGS__)```
```#define SHMEM_FUNC_PROTOTYPE(TYPE, OP, ...) shmem_ctx_##TYPE##_##OP(__VA_ARGS__)```
```#define SHMEM_FUNC_PROTOTYPE(TYPE, OP, ...) shmem_ctx_##TYPE##_atomic_##OP(__VA_ARGS__)```

Reply:
Yeah, I was thinking that too.  I'll give it a shot.

Reply:
I think this works, but `finc`/`fetch_inc` and `cswap`/`compare_swap` are exceptions, right?